### PR TITLE
[INTERN-13] Divide app into several modules./[INTERN-4] Cover CountriesApi with unit tests.

### PR DESCRIPTION
**Pull request by Strishwork**

### CODE REVIEW

File: api/build.gradle [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    }
+    kotlinOptions {
+        jvmTarget = '1.8'
+    }
+}
+
+dependencies {
+
+    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
+    implementation 'androidx.core:core-ktx:1.3.2'
```
Comment (yeoliinyk) [inline api/build.gradle]: Let's extract all the version numbers into variables inside project level build.gradle.

File: api/build.gradle [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    implementation 'androidx.appcompat:appcompat:1.2.0'
+    implementation 'com.google.android.material:material:1.2.1'
+    testImplementation 'junit:junit:4.+'
+    androidTestImplementation 'androidx.test.ext:junit:1.1.2'
+    androidTestImplementation 'androidx.test.espresso:espresso-core:3.3.0'
```
Comment (yeoliinyk) [inline api/build.gradle]: Why do we need espresso in the `api` module?

File: api/build.gradle [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        jvmTarget = '1.8'
+    }
+}
+
+dependencies {
+
+    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
+    implementation 'androidx.core:core-ktx:1.3.2'
+    implementation 'androidx.appcompat:appcompat:1.2.0'
+    implementation 'com.google.android.material:material:1.2.1'
```
Comment (yeoliinyk) [inline api/build.gradle]: Do we really need android dependencies in the `api` module?

File: api/build.gradle [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    compileSdkVersion 30
+    buildToolsVersion "30.0.2"
+
+    defaultConfig {
+        minSdkVersion 23
+        targetSdkVersion 30
+        versionCode 1
+        versionName "1.0"
+
+        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
```
Comment (yeoliinyk) [inline api/build.gradle]: Do we have android tests in the `api` module?

File: api/build.gradle (hunk 1 above)
Comment (Strishwork) [reply api/build.gradle]: Done.

File: api/build.gradle (hunk 2 above)
Comment (Strishwork) [reply api/build.gradle]: Removed.

File: api/build.gradle (hunk 3 above)
Comment (Strishwork) [reply api/build.gradle]: Removed.

File: api/build.gradle (hunk 4 above)
Comment (Strishwork) [reply api/build.gradle]: Removed.

File: api/build.gradle [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
+
+    implementation 'com.apollographql.apollo:apollo-runtime:2.4.1'
+    implementation 'com.apollographql.apollo:apollo-rx2-support:2.4.1'
+
+    testImplementation "org.mockito:mockito-inline:$mockito_version"
+    testImplementation "org.mockito:mockito-core:$mockito_version"
+    testImplementation "androidx.arch.core:core-testing:2.1.0"
+    testImplementation 'junit:junit:4.13.1'
```
Comment (yeoliinyk) [inline api/build.gradle]: I meant to extract all the version, not just Mockito :)

File: api/src/test/java/com/example/api/CountriesApiTest.kt [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import org.junit.Test
+import org.mockito.Answers
+import org.mockito.ArgumentCaptor
+import org.mockito.Mockito.*
+
+/**
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
```
Comment (yeoliinyk) [inline api/src/test/java/com/example/api/CountriesApiTest.kt]: Let's remove this.

File: api/src/test/java/com/example/api/CountriesApiTest.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+class CountriesApiTest {
+
+    @Rule
+    @JvmField
+    val instantExecutorRule = InstantTaskExecutorRule()
```
Comment (yeoliinyk) [inline api/src/test/java/com/example/api/CountriesApiTest.kt]: Why do we need this?

File: api/src/test/java/com/example/api/CountriesApiTest.kt [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    private val apolloClientMock = mock(ApolloClient::class.java, Answers.RETURNS_DEEP_STUBS)
+
+    @Test
+    fun test_apolloClientGetCountries() {
+        val countriesApi = spy(CountriesApi(apolloClientMock))
```
Comment (yeoliinyk) [inline api/src/test/java/com/example/api/CountriesApiTest.kt]: Usually, we shouldn't use `spy` for classes under tests. What was the motivation of doing this?

File: api/src/test/java/com/example/api/CountriesApiTest.kt [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+class CountriesApiTest {
+
+    @Rule
+    @JvmField
+    val instantExecutorRule = InstantTaskExecutorRule()
+
+    private val apolloClientMock = mock(ApolloClient::class.java, Answers.RETURNS_DEEP_STUBS)
```
Comment (yeoliinyk) [inline api/src/test/java/com/example/api/CountriesApiTest.kt:13]: 👍

File: app/build.gradle [hunk 10]
```diff
@@ -97,22 +90,13 @@ dependencies {
...
+
     debugImplementation 'androidx.fragment:fragment-testing:1.2.5'
-    androidTestImplementation('com.schibsted.spain:barista:3.7.0') {
-        exclude group: 'org.jetbrains.kotlin'
-    }
+    androidTestImplementation('com.schibsted.spain:barista:3.7.0')
     androidTestImplementation 'androidx.test:core-ktx:1.3.0'
     androidTestImplementation 'androidx.test.ext:junit-ktx:1.1.2'
 
     kaptAndroidTest "com.google.dagger:dagger-compiler:$dagger_version"
```
Comment (yeoliinyk) [inline app/build.gradle:98]: Same here, let's extract all the versions into the project level `build.gradle`.

File: api/build.gradle (hunk 5 above)
Comment (Strishwork) [reply api/build.gradle]: Done.

File: api/src/test/java/com/example/api/CountriesApiTest.kt (hunk 6 above)
Comment (Strishwork) [reply api/src/test/java/com/example/api/CountriesApiTest.kt]: Removed.

File: api/src/test/java/com/example/api/CountriesApiTest.kt (hunk 7 above)
Comment (Strishwork) [reply api/src/test/java/com/example/api/CountriesApiTest.kt]: Removed.

File: api/src/test/java/com/example/api/CountriesApiTest.kt (hunk 8 above)
Comment (Strishwork) [reply api/src/test/java/com/example/api/CountriesApiTest.kt]: Changed.

File: app/build.gradle (hunk 10 above)
Comment (Strishwork) [reply app/build.gradle:98]: Done.

**Review by yeoliinyk — APPROVED** [review]

**Pull request merged**